packages/backend/src/services/UnfurlService/UnfurlService.ts

```diff
@@ -0,0 +1,332 @@
+        browser = await puppeteer.connect({
+            browserWSEndpoint,
+        });
+
+        const page = await browser.newPage();
+
+        await page.setExtraHTTPHeaders({ cookie });
+
+        await page.setViewport({
+            width: 1400,
+            height: 768, // hardcoded
+        });
+
+        const blockedUrls = [
+            'headwayapp.co',
+            'rudderlabs.com',
+            'analytics.lightdash.com',
+            'cohere.so',
+            'intercom.io',
+        ];
+        await page.setRequestInterception(true);
+        page.on('request', (request: any) => {
+            const requestUrl = request.url();
+            if (blockedUrls.includes(requestUrl)) {
+                request.abort();
+                return;
+            }
+
+            request.continue();
+        });
+        await page.goto(parsedUrl.url, {
+            timeout: 100000,
+            waitUntil: 'networkidle0',
+        });
+        const path = `/tmp/${imageId}.png`;
+
+        const selector =
+            parsedUrl.lightdashPage === LightdashPage.DASHBOARD
+                ? '.react-grid-layout'
+                : `.echarts-for-react, [data-testid="visualization"]`; // Get .echarts-for-react, otherwise fallsback to data-testid (tables and bignumbers)
+        await page.waitForSelector(selector);
+        const element = await page.$(selector);
+
+        if (parsedUrl.lightdashPage === LightdashPage.DASHBOARD) {
+            // Remove navbar from screenshot
+            await page.evaluate((sel: any) => {
+                // @ts-ignore
+                const elements = document.querySelectorAll(sel);
+                elements.forEach((el) => el.parentNode.removeChild(el));
+            }, '.bp4-navbar');
+        }
+
+        const imageBuffer = await element.screenshot({
+            path,
+        });
+
+        return imageBuffer;
+    } catch (e) {
+        Logger.error(`Unable to fetch screenshots from headless chrome ${e}`);
+        return undefined;
+    } finally {
+        if (browser) await browser.close();
+    }
+};
+
+const getUserCookie = async (userUuid: string): Promise<string> => {
+    const token = getAuthenticationToken(userUuid);
+
+    const response = await fetch(
+        `${process.env.SITE_URL}/api/v1/headless-browser/login/${userUuid}`,
+        {
+            method: 'POST',
+            headers: {
+                'Content-Type': 'application/json',
+            },
+            body: JSON.stringify({ token }),
+        },
+    );
+    if (response.status !== 200) {
+        throw new Error(
+            `Unable to get cookie for user ${userUuid}: ${await response.text()}`,
+        );
+    }
+    const header = response.headers.get('set-cookie');
+    if (header === null) {
+        const loginBody = await response.json();
+        throw new AuthorizationError(
+            `Cannot sign in:\n${JSON.stringify(loginBody)}`,
+        );
+    }
+    return header;
+};
```

Comment:
I feel like this file does too many things at once. maybe split out saveScreenshot and getUserCookie as a separate service. because I was expecting UnfurlService to only export Unfurl class. or maybe its part of the unfurlService because I see a reference there.

Reply:
ok since its not exported anywhere, maybe a private static method